### ISSUE TITLE
How’s the cold start time of EC2 with instance store

### ISSUE THREAD

**Issue by Weixuanf** [aws-samples/comfyui-on-eks#9]
Cuz I want to scale down to 0 instance when there’s no request.

How long does it take to cold start EC2 from 0 instance? I think instance store EC2 is slower to boot than EBS backed EC2 instance?

and downloading models from S3 to instance store takes extra time. How’s the downloading speed look like from s3 to instance store?

Thanks for this amazing template!

**Comment by Shellmode** [comment]
It might take several minutes to ten minutes to cold start, here are some steps:
1. Karpenter finds provisionable pod(s), starts to spin up EC2(negligible time consumption, less than 1s)
2. EC2 initialize, run user-data script(sync all models on S3 to instance store) defined in [Karpenter custom resource EC2NodeClass](https://github.com/aws-samples/comfyui-on-eks/blob/main/manifests/Karpenter/karpenter_v1beta1.yaml), start kubelet and other stuffs to get node ready. (It may take a few minutes, depends on how much you need to sync from S3)
3. Pull image from ECR(It may also take like 5mins, depends on how large the image is)

Actually all GPU instance types (like [g4dn](https://aws.amazon.com/ec2/instance-types/g4/)/[g5](https://aws.amazon.com/ec2/instance-types/g5/)/[g6](https://aws.amazon.com/ec2/instance-types/g6/)) have instance store, boot time is the same whether you use instance store or not. 

The solution uses instance store to improve models loading and switching (inside ComfyUI) performance. 

It's a tradeoff, we spend more time setting up the environment to get better performance.

**Comment by Weixuanf** [comment]
thanks very much for your reply. I want to run serverless comfyui servers that scale down to 0 when no requests to save time. So cold start time is very important for me. I hope to get < 5s cold start time (excluding comfyui boot time itself). I'm thinking EC2 + EBS and stopping/starting the EC2 server to achieve better cold start times than using auto scale group. If you have other suggestions, please let me know!

> Actually all GPU instance types (like [g4dn](https://aws.amazon.com/ec2/instance-types/g4/)/[g5](https://aws.amazon.com/ec2/instance-types/g5/)/[g6](https://aws.amazon.com/ec2/instance-types/g6/)) have instance store, boot time is the same whether you use instance store or not.

oh so even if I use EKS, there will still be instance store in it?

**Comment by Shellmode** [comment]
Yes, g4dn & g5 & g6 all have instance store, refer to [Amazon EC2 instance store](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html), you can use it or just ignore it (it's free).

EC2 with EBS will have less boot time, because there's no image pulling and model syncing. But you need to handle EC2 scaling in/out yourself. Besides that, loading models from EBS to GPU memory might take more time than loading from instance store.

**Comment by PeterTF656** [comment]
> thanks very much for your reply. I want to run serverless comfyui servers that scale down to 0 when no requests to save time. So cold start time is very important for me. I hope to get < 5s cold start time (excluding comfyui boot time itself). I'm thinking EC2 + EBS and stopping/starting the EC2 server to achieve better cold start times than using auto scale group. If you have other suggestions, please let me know!
> 
> > Actually all GPU instance types (like [g4dn](https://aws.amazon.com/ec2/instance-types/g4/)/[g5](https://aws.amazon.com/ec2/instance-types/g5/)/[g6](https://aws.amazon.com/ec2/instance-types/g6/)) have instance store, boot time is the same whether you use instance store or not.
> 
> oh so even if I use EKS, there will still be instance store in it?

Have you thought about mounting EFS to your instances?

**Comment by JTipecska** [comment]
Great question — a few concrete numbers and an update for the current template:

Cold start from zero is minutes, not seconds, and breaks down as: Karpenter provisioning the EC2 (~seconds) → node init + user-data model sync from S3 to instance store (a few minutes, depends on model size) → ECR image pull (can be ~5 min for a large image). All g4dn/g5/g6 instances include instance store; boot time is the same whether you use it or not — the template uses it purely to speed model loading/switching inside ComfyUI.

Scale-to-zero itself works on the current template — Karpenter v1 with consolidationPolicy: WhenEmptyOrUnderutilized + expireAfter: 168h removes idle GPU nodes (scale the ComfyUI Deployment to 0 replicas, or drive it with a request-based autoscaler, to reach true zero).

Your <5s target is not achievable with this EKS + Karpenter + instance-store design — the model sync and image pull dominate. To get near-instant starts you'd need warm capacity or a different pattern: keep a minimum of 1 GPU node (defeats scale-to-zero), pre-bake models into a large AMI/EBS snapshot (Shellmode's EBS point — faster boot, but you manage scaling and model loading is slower from EBS), mount EFS/FSx for shared model storage (PeterTF656's suggestion — avoids per-node S3 sync but adds network-load latency), or use a snapshot/warm-pool approach outside this template. Net: this template optimizes throughput and cost (scale-to-zero, fast model switching once warm), not sub-5s cold start.

Since this is a design/Q&A thread and scale-to-zero now works, closing as answered — reopen if you want to dig into a specific low-latency pattern.
